### PR TITLE
use named tuples as keyword varargs. fixes #4916

### LINKED ISSUE #4916: variable numbers of keyword arguments as dict
``` julia
foo(; args...)
```

Inside of a function with this signature, `args` will be an array of tuples containing symbols and values. Cool, but the array will be unordered, so I pretty much have to loop over it trying to match symbols.

I think it would be really cool if `args` came in as a dictionary. I'm guessing there would be some overhead involved in that, and that's perhaps why it has not been done. But I wonder if the overhead involved in looping over symbols and doing a bunch of string comparisons (I think that's what's happening, could be wrong) might even outweigh the dict creation.

I didn't see an issue, but has this been considered?

### LINKED ISSUE #9972: Better error needed for misuse of keyword args
I had a mental lapse and forgot to use the keyword argument name when invoking a function. I ended being baffled for a good five minutes before realizing what happened.

```
julia> f(;x=1) = x+1
f (generic function with 1 method)

julia> f(;x=3)
4

julia> f(;3)
ERROR: BoundsError
 in anonymous at no file
```

This seems like something the parser should be able to pick up instantly.

### PR DESCRIPTION
**Pull request by JeffBezanson**
Same highly-biased benchmark as #21915 (~100x speedup):

```
julia> f(;kw...) = g(;kw...);

julia> g(;kw...) = kw;

julia> c = g(a=1,b=2);

julia> @time for i = 1:10000; f(a=1,b=2); end
  0.000010 seconds

julia> @time for i = 1:10000; f(;c...); end
  0.000577 seconds (20.00 k allocations: 625.000 KiB)
```

This also helps a lot towards #9551. The PR makes the lowering of keyword argument methods simpler and easier to optimize (note the net code deletion). In fact I think just a bit of constant propagation (#24362) on top of this will solve the problem entirely.

Also fixes #9972.

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -14,6 +14,10 @@ New language features
   * Named tuples, with the syntax `(a=1, b=2)`. These behave very similarly to tuples,
     except components can also be accessed by name using dot syntax `t.a` ([#22194]).
 
+  * Keyword argument containers (`kw` in `f(; kw...)`) are now named tuples. Dictionary
+    functions like `haskey` and indexing can be used on them, and name-value pairs can be
+    iterated using `pairs(kw)` ([#4916]).
+
```
Comment (quinnj) [inline NEWS.md]: I wonder if it'd be worth mentioning that keyword arguments can no longer be repeated; I know that wasn't changed explicitly in this PR, but seems like a good place to mention it anyway.

**Comment by vtjnash** [comment]
What constant propagation would that be? Currently that PR can’t really do anything for NamedTuples (which would likely need more of a MemSSA or GVN pass).

**Comment by JeffBezanson** [comment]
The code here calls `haskey` to see whether a keyword argument was passed or should get its default value. If we can remove the branch then code with keyword arguments can be fully inferred. In fact it already works in many cases, if the passed value has the same type as the default value.

**Comment by ChrisRackauckas** [comment]
What's the timing now that https://github.com/JuliaLang/julia/pull/24362 merged?

**Comment by SimonDanisch** [comment]
```Julia
julia> f(;kw...) = g(;kw...);

julia> g(;kw...) = kw;

julia> c = g(a=1,b=2);

julia> @time for i = 1:10000; f(a=1,b=2); end
  0.000008 seconds
julia> @time for i = 1:10000; f(;c...); end
  0.000377 seconds (20.00 k allocations: 625.000 KiB)
```
This has a bit of noise in execution time though...

**Comment by yuyichao** [comment]
It's not measuring anything...

```julia
julia> f(;kw...) = g(;kw...)
f (generic function with 1 method)

julia> g(;kw...) = kw
g (generic function with 1 method)

julia> const c = g(a=1, b=2)
(a = 1, b = 2)

julia> @time for i in 1:1000; f(a=1, b=2); end
  0.000003 seconds

julia> @time for i in 1:1000; f(;c...); end
  0.000003 seconds
```

**Pull request merged**